File: EstructuradedatosU2java/src/actividad2unidad2/Estaticas.java
package actividad2unidad2;

import java.util.Arrays;

public class Estaticas {
    public static void main(String[] args) {
        int[] array = {64, 25, 12, 22, 11};
        
        // Ordenamiento Burbuja
        bubbleSort(array);
        
        // Ordenamiento Mezcla
        mergeSort(array);
        
        // Ordenamiento Cocktail
        cocktailSort(array);
        
        // Ordenamiento Inserción
        insertionSort(array);
        
        // Ordenamiento Selección
        selectionSort(array);
    }
    
    // Método para el Ordenamiento Burbuja
    public static void bubbleSort(int[] array) {
        int n = array.length;
        int[] sortedArray = Arrays.copyOf(array, n);
        for (int i = 0; i < n-1; i++) {
            for (int j = 0; j < n-i-1; j++) {
                if (sortedArray[j] > sortedArray[j+1]) {
                    int temp = sortedArray[j];
                    sortedArray[j] = sortedArray[j+1];
                    sortedArray[j+1] = temp;
                }
            }
        }
        System.out.println("Bubble Sort: " + Arrays.toString(sortedArray));
    }
    
    // Método para el Ordenamiento por Mezcla
    public static void mergeSort(int[] array) {
        int n = array.length;
        if (n < 2) {
            return;
        }
        int middle = n / 2;
        int[] leftArray = new int[middle];
        int[] rightArray = new int[n - middle];
        
        // Dividir el array en dos sub-arrays
        for (int i = 0; i < middle; i++) {
            leftArray[i] = array[i];
        }
        for (int i = middle; i < n; i++) {
            rightArray[i - middle] = array[i];
        }
        
        // Llamada recursiva para ordenar los sub-arrays
        mergeSort(leftArray);
        mergeSort(rightArray);
        
        // Combinar los sub-arrays ordenados
        merge(array, leftArray, rightArray);
    }
    
    // Combinar dos sub-arrays ordenados
    public static void merge(int[] array, int[] leftArray, int[] rightArray) {
        int n1 = leftArray.length;
        int n2 = rightArray.length;
        int i = 0, j = 0, k = 0;
        while (i < n1 && j < n2) {
            if (leftArray[i] <= rightArray[j]) {
                array[k] = leftArray[i];
                i++;
            } else {
                array[k] = rightArray[j];
                j++;
            }
            k++;
        }
        while (i < n1) {
            array[k] = leftArray[i];
            i++;
            k++;
        }
        while (j < n2) {
            array[k] = rightArray[j];
            j++;
            k++;
        }
        System.out.println("Merge Sort: " + Arrays.toString(array));
    }
    
    // Método para el Ordenamiento Cocktail
    public static void cocktailSort(int[] array) {
        int n = array.length;
        boolean swapped = true;
        int start = 0;
        int end = n - 1;
        
        while (swapped) {
            swapped = false;
            
            // Mueve el elemento más grande al final
            for (int i = start; i < end; ++i) {
                if (array[i] > array[i + 1]) {
                    int temp = array[i];
                    array[i] = array[i + 1];
                    array[i + 1] = temp;
                    swapped = true;
                }
            }
            
            // SI nada se mueve, se ordena la matriz
            if (!swapped)
                break;
            
            swapped = false;
            end--;
            
            // Mueve el elemento más pequeño al inicio
            for (int i = end - 1; i >= start; --i) {
                if (array[i] > array[i + 1]) {
                    int temp = array[i];
                    array[i] = array[i + 1];
                    array[i + 1] = temp;
                    swapped = true;
                }
            }
            
            start++;
        }
        
        System.out.println("Cocktail Sort: " + Arrays.toString(array));
    }
    
    // Método para el Ordenamiento por inserción
    public static void insertionSort(int[] array) {
        int n = array.length;
        for (int i = 1; i < n; ++i) {
            int key = array[i];
            int j = i - 1;
            while (j >= 0 && array[j] > key) {
                array[j + 1] = array[j];
                j = j - 1;
            }
            array[j + 1] = key;
        }
        System.out.println("Insertion Sort: " + Arrays.toString(array));
    }
    
    // Método para el Ordenamiento por selección
    public static void selectionSort(int[] array) {
        int n = array.length;
        for (int i = 0; i < n-1; i++) {
            int minIndex = i;
            for (int j = i+1; j < n; j++) {
                if (array[j] < array[minIndex]) {
                    minIndex = j;
                }
            }
            // Se cambia aquí el elemento mínimo encontrado con el primer elemento
            int temp = array[minIndex];
            array[minIndex] = array[i];
            array[i] = temp;
        }
        System.out.println("Selection Sort: " + Arrays.toString(array));
    }
}

File: EstructuradedatosU2py/Estaticas.py
def Estaticas():
    #Se indica a la matriz para que ordene en tipo burbuja
   def bubbleSort(array):
    n = len(array)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]
#Se indica a la matriz para que ordene en tipo mezcla
def mergeSort(array):
    if len(array) > 1:
        middle = len(array) // 2
        left_half = array[:middle]
        right_half = array[middle:]

        mergeSort(left_half)
        mergeSort(right_half)

        i = j = k = 0

        while i < len(left_half) and j < len(right_half):
            if left_half[i] < right_half[j]:
                array[k] = left_half[i]
                i += 1
            else:
                array[k] = right_half[j]
                j += 1
            k += 1

        while i < len(left_half):
            array[k] = left_half[i]
            i += 1
            k += 1

        while j < len(right_half):
            array[k] = right_half[j]
            j += 1
            k += 1
#Se indica a la matriz para que ordene en tipo cocktail
def cocktailSort(array):
    n = len(array)
    swapped = True
    start = 0
    end = n - 1
    
    while swapped:
        swapped = False
        
        #Este ciclo mueve el elemento más grande al final
        for i in range(start, end):
            if array[i] > array[i + 1]:
                array[i], array[i + 1] = array[i + 1], array[i]
                swapped = True
        
        # El condicional si no se mueve nada, entonces se ordena la matriz
        if not swapped:
            break
        
        swapped = False
        end -= 1
        
        #Se mueve el elemento más pequeño al inicio
        for i in range(end - 1, start - 1, -1):
            if array[i] > array[i + 1]:
                array[i], array[i + 1] = array[i + 1], array[i]
                swapped = True
        
        start += 1
#Se indica a la matriz para que ordene en tipo inserción
def insertionSort(array):
    for i in range(1, len(array)):
        key = array[i]
        j = i - 1
        while j >= 0 and array[j] > key:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = key
#Se indica a la matriz para que ordene en tipo selección
def selectionSort(array):
    n = len(array)
    for i in range(n - 1):
        min_index = i
        for j in range(i + 1, n):
            if array[j] < array[min_index]:
                min_index = j
        array[i], array[min_index] = array[min_index], array[i]
#Se indica el orden impresión de las matrices
def printArray(array):
    print("Ordenamiento por Burbuja:", ", ".join(map(str, array)))
    print("Ordenamiento por Mezcla:", ", ".join(map(str, array)))
    print("Ordenamiento por Cocktail:", ", ".join(map(str, array)))
    print("Ordenamiento por Inserción:", ", ".join(map(str, array)))
    print("Ordenamiento por Selección:", ", ".join(map(str, array)))

array = [64, 25, 12, 22, 11]

#bubbleSort(array) 
mergeSort(array)
cocktailSort(array)
insertionSort(array)
selectionSort(array)

printArray(array)

if __name__ == "__main__":
    Estaticas()
